### PR TITLE
RFD: Moderated File Transfers (web)

### PR DESCRIPTION
**Pull request by avatus**
Rendered view [here](https://github.com/gravitational/teleport/blob/ad600b4359340491db6e11d792cabf2e8c2cc212/rfd/0113-moderated-file-transfers.md)
Issue https://github.com/gravitational/teleport/issues/16008

### CODE REVIEW

File: rfd/0113-moderated-file-transfers.md
Comment (capnspacehook) [inline rfd/0113-moderated-file-transfers.md:11]: Does the security and/or product team need to approve this as well?

File: rfd/0113-moderated-file-transfers.md
Comment (capnspacehook) [inline rfd/0113-moderated-file-transfers.md:175]: I agree, I'd personally lean towards option 1 as well.

File: rfd/0113-moderated-file-transfers.md
Comment (capnspacehook) [inline rfd/0113-moderated-file-transfers.md:48]: If this was abstracted, this could allow access requests for a single command, instead of an interactive session (if that isn't already possible), would be interesting

File: rfd/0113-moderated-file-transfers.md
Comment (avatus) [reply rfd/0113-moderated-file-transfers.md:11]: Yeah, wasn't sure who to add. Thanks for the nudge.

**Review by xacrimon — COMMENTED** [review]
first pass

File: rfd/0113-moderated-file-transfers.md
Comment (avatus) [reply rfd/0113-moderated-file-transfers.md]: Even better, I just replaced the "api server" with proxy since thats what it is anyway. Moved it closer to the requester entity as well.

File: rfd/0113-moderated-file-transfers.md
Comment (avatus) [reply rfd/0113-moderated-file-transfers.md:63]: Will update soon. I'm not sure we need to store the entire party yet, but yes, it would be "anyone who is peer or moderator"

File: rfd/0113-moderated-file-transfers.md
Comment (avatus) [reply rfd/0113-moderated-file-transfers.md:60]: In this context, I would use `scx.Identity.TeleportUser`... however, maybe a more contextual name would be `requestingTeleportUser`. Just so it's very clear its the requester and what "kind" of user.

Comment (xacrimon) [reply rfd/0113-moderated-file-transfers.md:60]: Yep, a more contextual name is what I wanted. TBH, calling it `requester` is fine and avoids annoying name size blowup, just throw a comment on the field saying it's a Teleport user so it's obvious on hover.

File: rfd/0113-moderated-file-transfers.md
Comment (zmb3) [inline rfd/0113-moderated-file-transfers.md]: Are we leaving tsh out of scope because it requires a different solution or because we just haven't thought about it yet?

From a design perspective, I think it's important to understand why we're proposing a solution that's only valid for web.

File: rfd/0113-moderated-file-transfers.md
Comment (zmb3) [inline rfd/0113-moderated-file-transfers.md:83]: Is `ssh.Request` and existing type? This is the first mention of it.

File: rfd/0113-moderated-file-transfers.md
Comment (zmb3) [inline rfd/0113-moderated-file-transfers.md]: Will this new request only be used for moderated sessions or will it apply any time there is a web based file transfer? 

If it won't apply every time, how will you detect whether or not you need to use it?

File: rfd/0113-moderated-file-transfers.md
Comment (zmb3) [inline rfd/0113-moderated-file-transfers.md:155]: `request` is a vague term and makes me think of access requests here. Maybe `isApprovedFileTransfer` or something along those lines would be a better name.

File: rfd/0113-moderated-file-transfers.md
Comment (zmb3) [inline rfd/0113-moderated-file-transfers.md:183]: It does need to propagate to all proxies though, right?

File: rfd/0113-moderated-file-transfers.md
Comment (avatus) [reply rfd/0113-moderated-file-transfers.md]: I believe the backend portion of this solution could be reused for `tsh` quite easily. I don't have the context yet to decide what the UI/UX would look like for `tsh`. I can do more research and revisit.

File: rfd/0113-moderated-file-transfers.md
Comment (avatus) [reply rfd/0113-moderated-file-transfers.md]: Only for moderated sessions. If it isn't a moderated session, no extra parameters are needed to be sent and will just pass the request without them added. I speak about [it here](https://github.com/gravitational/teleport/blob/ccdd62aa2d94e25dd896591db4057fec73359962/rfd/0113-moderated-file-transfers.md#updated-file-transfer-api-handler) but I see its a bit unclear now. Will update.

The web console will determine if we need to start this new process, and the api handler knowing when to add the environment variables based on the existence of the optional parameters sent. Sort of similar to how we [issue certs when a webauthn response is included](https://github.com/gravitational/teleport/pull/22864/files#diff-ad4065e07d2573425dc5f7460e26dd324ffd893845ae1d3fe4b997113aba90cbR133), we would be able to set the environment variables needed for `OpenExecSession`. 

Also, `OpenExecSession` knows if the soon-to-be-opened session is moderated or not, so we can have a check "if moderated, see if they've sent through a commandRequestID".

File: rfd/0113-moderated-file-transfers.md
Comment (avatus) [reply rfd/0113-moderated-file-transfers.md:183]: The `ApprovedFileTransferRequest` struct only needs to live in the node memory and doesn't need to be propagated to the proxies. the proxy in this case is only used to send the request/approval through the ssh channel connected to the proxy for each member.

File: rfd/0113-moderated-file-transfers.md
Comment (avatus) [reply rfd/0113-moderated-file-transfers.md:155]: `isApprovedFileTransfer` will work great. If we want, we can even abstract it early (and in other places in the code) to `isApprovedCommand`? its only being implemented with file transfers right now, this process of requesting an command, approving a command, and running that command seems like it can be related to anything we want in the future. Thoughts?

File: rfd/0113-moderated-file-transfers.md
Comment (avatus) [reply rfd/0113-moderated-file-transfers.md]: Commenting a few more notes/thinking out loud. 

With the web, the request is outside of the shell and handled via the UI, as well as the approval process. This means the shell isn't blocked by any file transfer requests. 

For cli, we can maybe inform the approvers in the cli that someone is requesting a file transfer and implement an approve/deny command, similar to the force-terminate command. I'm not sure how this would scale will multiple file transfer requests at once, but its a thread to pull on.

I think the `tsh scp` command could function after approval similar to the new file transfer http request optional params by adding two new flags, the `--sessionId` and `--commandRequestId`.

 I'll continue to ponder.

File: rfd/0113-moderated-file-transfers.md
Comment (zmb3) [inline rfd/0113-moderated-file-transfers.md:186]: I think it's worth discussing the UX for `tsh`-based file transfers. Will it work? Will it fail with a confusing error? Can we detect it and prompt the user to initiate the transfer via the web UI?

**Comment by zmb3** [comment]
Hey @xacrimon and @capnspacehook - I think all of your comments have been addressed so far.

There's still a few corner cases we're experimenting with, so we'll probably keep this open until we decide on a a path forward, but I wanted to confirm with you two if there's anything else you'd like to see addressed before we start pushing more PRs.

**Comment by capnspacehook** [comment]
LGTM as long as the flow with `tsh scp` gets addressed eventually

**Comment by zmb3** [comment]
> LGTM as long as the flow with `tsh scp` gets addressed eventually

To be honest, we might not ever support `tsh scp`, as the whole idea is you have a moderated _session_, where moderators are already present and watching. Technically an SCP is a separate session, but in the web UI we can make it appear as part of an existing session. We are unsure how this could work with `tsh`-initiated transfers.

@avatus it would be worth considering whether we can allow moderators to approve file transfers if they joined via `tsh join`, so long as the user is in a web UI session. Let's add a bit more info about this to the RFD.

**Comment by xacrimon** [comment]
Hey, everything to me looks fine so far and I'd be happy to see it move forward. Feel free to hit `Request Review` once the changes @zmb3 mentioned are added.

**Comment by zmb3** [comment]
In the future, please try to follow the branch naming conventions for RFDs: https://github.com/gravitational/teleport/blob/785fa046276cee7adde54a7c5cc31b013cc7a65f/rfd/0000-rfds.md?plain=1#L68-L70

**Comment by avatus** [comment]
Updated to `implemented` as well as the RFD number to what I believe was the next free code. Will follow the branch naming conventions next time 👍

**Comment by avatus** [comment]
Friendly ping @zmb3 @capnspacehook @xacrimon

**Review by zmb3 — APPROVED** [review]

**Review by russjones — APPROVED** [review]

**Pull request merged**
